jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -234,30 +234,29 @@ protected void doStop() throws Exception
         // If stop timeout try to gracefully stop
         long timeout = getStopTimeout();
         BlockingQueue<Runnable> jobs = getQueue();
-        if (timeout > 0)
-        {
-            // Fill the job queue with noop jobs to wakeup idle threads.
-            for (int i = 0; i < threads; ++i)
-            {
-                jobs.offer(NOOP);
-            }
 
-            // try to let jobs complete naturally for half our stop time
-            joinThreads(System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout) / 2);
+        // Fill the job queue with noop jobs to wakeup idle threads.
+        for (int i = 0; i < threads; ++i)
+            jobs.offer(NOOP);
 
-            // If we still have threads running, get a bit more aggressive
+        // If we have a timeout, try to let jobs complete naturally for half our stop time
+        if (timeout > 0)
+            joinThreads(System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeout) / 2);
+        Thread.yield();
 
-            // interrupt remaining threads
-            for (Thread thread : _threads)
-            {
-                if (LOG.isDebugEnabled())
-                    LOG.debug("Interrupting {}", thread);
-                thread.interrupt();
-            }
+        // If we still have threads running, get a bit more aggressive
+        // interrupt remaining threads
+        for (Thread thread : _threads)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Interrupting {}", thread);
+            thread.interrupt();
+        }
```

Comment:
Now it may be possible that we interrupt threads immediately after submitting the NOOP jobs (or normal jobs) -- I'm not sure we want that if we want to gracefully stop (by letting the current tasks finish).

Also, interrupting is the right thing to do, but e.g. `ReservedThreadExecutor` does not handle it well: it just loops around and goes again into `take()/poll()`.

Reply:
Ah that is a problem. The `ReservedThread#run` method does do a `while(isRunning)`, but there is an inner loop in `ReservedThread#reservedWait` that doesn't.    I'll fix that....

Reply:
last check in fixed the inner loop of the ReservedThread